src/debugger/gdb-stub.c

```diff
@@ -433,7 +440,52 @@ static void _processQSupportedCommand(struct GDBStub* stub, const char* message)
 		}
 		message = end + 1;
 	}
-	strncpy(stub->outgoing, "swbreak+;hwbreak+", GDB_STUB_MAX_LINE - 4);
+	strncpy(stub->outgoing, "swbreak+;hwbreak+;qXfer:features:read+;qXfer:memory-map:read+", GDB_STUB_MAX_LINE - 4);
+}
+
+static void _processQXferCommand(struct GDBStub* stub, const char* params, const char* data) {
+	unsigned offset = 0;
+	unsigned length = 0;
+
+	unsigned index = 0;
+	for (index = 0; params[index] != ','; ++index) {
+		offset <<= 4;
+		offset |= _hex2int(&params[index], 1);
+	}
+
+	++index;
+	for (; index < strlen(params) - 3; ++index)
+	{
+		length <<= 4;
+		length |= _hex2int(&params[index], 1);
+	}
+
+	memset(stub->outgoing, 0, GDB_STUB_MAX_LINE);
+	if (strlen(data) < length + offset) {
+		length = strlen(data) - offset;
+		stub->outgoing[0] = 'l';
+	} else {
+		stub->outgoing[0] = 'm';
+	}
+	strncpy(&stub->outgoing[1], &data[offset], length);
+}
+
+static void _generateMemoryMapXml(struct GDBStub* stub, char* memoryMap) {
+	int index = 0;
+	strncpy(memoryMap, "<memory-map version=\"1.0\">", 27);
+	index += 26;
+	const struct mCoreMemoryBlock* blocks;
+	size_t nBlocks = stub->d.core->listMemoryBlocks(stub->d.core, &blocks);
+	size_t i;
+	for (i = 0; i < nBlocks; ++i) {
+		if (!(blocks[i].flags & mCORE_MEMORY_MAPPED))
+			continue;
+
+		char* type = blocks[i].flags & mCORE_MEMORY_WRITE ? "ram" : "rom";
+		snprintf(&memoryMap[index], 60, "<memory type=\"%s\" start=\"0x%08x\" length=\"0x%08x\"/>", type, blocks[i].start, blocks[i].size);
```

Comment:
This is not at all safe and not how you should be doing this, The size  you should be passing is the size remaining in the buffer, not the size of the string you're copying...